File: mock-api-server/src/controllers/requirementCompletionController.ts
import { Request, Response } from "express";

import { mockProjectRequirements } from "../data/mockProjectRequirements.js";
import { logger } from "../utils/logger.js";
import { mockDelayHandler } from "../utils/mockDelay.js";

export class RequirementCompletionController {
  /**
   * Get requirement completion time analytics
   */
  async getCompletionAnalytics(req: Request, res: Response) {
    await mockDelayHandler();

    try {
      logger.info("Fetching requirement completion analytics");

      const now = new Date();
      const requirements = mockProjectRequirements;

      // Calculate completion metrics
      const completedRequirements = requirements.filter(
        (req) => req.status === "completed",
      );
      const inDevelopmentRequirements = requirements.filter(
        (req) => req.status === "in-development",
      );
      const approvedRequirements = requirements.filter(
        (req) => req.status === "approved",
      );

      // On-time completion analysis
      const onTimeCompleted = completedRequirements.filter((req) => {
        const expectedDate = new Date(req.expectedCompletionDate);
        const updatedDate = new Date(req.updatedAt);

        return updatedDate <= expectedDate;
      });

      // Late completion analysis
      const lateCompleted = completedRequirements.filter((req) => {
        const expectedDate = new Date(req.expectedCompletionDate);
        const updatedDate = new Date(req.updatedAt);

        return updatedDate > expectedDate;
      });

      // Overdue (still in development past expected date)
      const overdueRequirements = inDevelopmentRequirements.filter((req) => {
        const expectedDate = new Date(req.expectedCompletionDate);

        return now > expectedDate;
      });

      // At risk (approaching deadline within 7 days)
      const atRiskRequirements = [
        ...inDevelopmentRequirements,
        ...approvedRequirements,
      ].filter((req) => {
        const expectedDate = new Date(req.expectedCompletionDate);
        const daysUntilDeadline = Math.ceil(
          (expectedDate.getTime() - now.getTime()) / (1000 * 60 * 60 * 24),
        );

        return daysUntilDeadline <= 7 && daysUntilDeadline > 0;
      });

      // Calculate average completion time
      const avgCompletionTime =
        lateCompleted.length > 0
          ? lateCompleted.reduce((sum, req) => {
              const expectedDate = new Date(req.expectedCompletionDate);
              const completedDate = new Date(req.updatedAt);
              const daysLate = Math.ceil(
                (completedDate.getTime() - expectedDate.getTime()) /
                  (1000 * 60 * 60 * 24),
              );

              return sum + daysLate;
            }, 0) / lateCompleted.length
          : 0;

      // Priority breakdown
      const byPriority = {
        high: {
          total: requirements.filter((req) => req.priority === "high").length,
          onTime: onTimeCompleted.filter((req) => req.priority === "high")
            .length,
          late: lateCompleted.filter((req) => req.priority === "high").length,
          overdue: overdueRequirements.filter((req) => req.priority === "high")
            .length,
        },
        medium: {
          total: requirements.filter((req) => req.priority === "medium").length,
          onTime: onTimeCompleted.filter((req) => req.priority === "medium")
            .length,
          late: lateCompleted.filter((req) => req.priority === "medium").length,
          overdue: overdueRequirements.filter(
            (req) => req.priority === "medium",
          ).length,
        },
        low: {
          total: requirements.filter((req) => req.priority === "low").length,
          onTime: onTimeCompleted.filter((req) => req.priority === "low")
            .length,
          late: lateCompleted.filter((req) => req.priority === "low").length,
          overdue: overdueRequirements.filter((req) => req.priority === "low")
            .length,
        },
      };

      // Monthly trend (last 6 months)
      const monthlyTrend = [];

      for (let i = 5; i >= 0; i--) {
        const monthDate = new Date(now.getFullYear(), now.getMonth() - i, 1);
        const nextMonthDate = new Date(
          now.getFullYear(),
          now.getMonth() - i + 1,
          1,
        );

        const monthRequirements = completedRequirements.filter((req) => {
          const completedDate = new Date(req.updatedAt);

          return completedDate >= monthDate && completedDate < nextMonthDate;
        });

        const monthOnTime = monthRequirements.filter((req) => {
          const expectedDate = new Date(req.expectedCompletionDate);
          const completedDate = new Date(req.updatedAt);

          return completedDate <= expectedDate;
        });

        monthlyTrend.push({
          month: monthDate.toISOString().slice(0, 7), // YYYY-MM format
          total: monthRequirements.length,
          onTime: monthOnTime.length,
          onTimeRate:
            monthRequirements.length > 0
              ? Math.round(
                  (monthOnTime.length / monthRequirements.length) * 100,
                )
              : 0,
        });
      }

      const analytics = {
        summary: {
          totalRequirements: requirements.length,
          completedRequirements: completedRequirements.length,
          onTimeCompleted: onTimeCompleted.length,
          lateCompleted: lateCompleted.length,
          overdueRequirements: overdueRequirements.length,
          atRiskRequirements: atRiskRequirements.length,
          onTimeRate:
            completedRequirements.length > 0
              ? Math.round(
                  (onTimeCompleted.length / completedRequirements.length) * 100,
                )
              : 0,
          avgDelayDays: Math.round(avgCompletionTime * 10) / 10, // Round to 1 decimal
        },
        byPriority,
        monthlyTrend,
        overdueItems: overdueRequirements.map((req) => ({
          id: req.id,
          name: req.name,
          priority: req.priority,
          expectedDate: req.expectedCompletionDate,
          daysOverdue: Math.ceil(
            (now.getTime() - new Date(req.expectedCompletionDate).getTime()) /
              (1000 * 60 * 60 * 24),
          ),
          assignedAnalyst: req.assignedAnalyst,
          projectName: req.project?.applicationName,
        })),
        atRiskItems: atRiskRequirements.map((req) => ({
          id: req.id,
          name: req.name,
          priority: req.priority,
          expectedDate: req.expectedCompletionDate,
          daysUntilDeadline: Math.ceil(
            (new Date(req.expectedCompletionDate).getTime() - now.getTime()) /
              (1000 * 60 * 60 * 24),
          ),
          assignedAnalyst: req.assignedAnalyst,
          projectName: req.project?.applicationName,
        })),
      };

      res.status(200).json({
        success: true,
        data: analytics,
      });
    } catch (error) {
      logger.error("Error fetching completion analytics:", error);
      res.status(500).json({
        success: false,
        error: {
          message: "Failed to fetch completion analytics",
          code: "COMPLETION_ANALYTICS_ERROR",
        },
      });
    }
  }

  /**
   * Get detailed completion metrics for a specific time period
   */
  async getCompletionMetrics(req: Request, res: Response) {
    await mockDelayHandler();

    try {
      const { period = "month", analystId } = req.query;

      logger.info(
        `Fetching completion metrics for period: ${period}, analyst: ${analystId}`,
      );

      let requirements = mockProjectRequirements;

      // Filter by analyst if specified
      if (analystId) {
        requirements = requirements.filter(
          (req) => req.assignedAnalyst === parseInt(analystId as string),
        );
      }

      const now = new Date();
      let startDate: Date;

      // Determine date range based on period
      switch (period) {
        case "week":
          startDate = new Date(
            now.getFullYear(),
            now.getMonth(),
            now.getDate() - 7,
          );
          break;
        case "quarter":
          startDate = new Date(
            now.getFullYear(),
            now.getMonth() - 3,
            now.getDate(),
          );
          break;
        case "year":
          startDate = new Date(
            now.getFullYear() - 1,
            now.getMonth(),
            now.getDate(),
          );
          break;
        default: // month
          startDate = new Date(
            now.getFullYear(),
            now.getMonth() - 1,
            now.getDate(),
          );
      }

      // Filter requirements by time period
      const periodRequirements = requirements.filter((req) => {
        const updatedDate = new Date(req.updatedAt);

        return updatedDate >= startDate;
      });

      const completedInPeriod = periodRequirements.filter(
        (req) => req.status === "completed",
      );
      const onTimeInPeriod = completedInPeriod.filter((req) => {
        const expectedDate = new Date(req.expectedCompletionDate);
        const completedDate = new Date(req.updatedAt);

        return completedDate <= expectedDate;
      });

      const metrics = {
        period,
        startDate: startDate.toISOString(),
        endDate: now.toISOString(),
        totalRequirements: periodRequirements.length,
        completedRequirements: completedInPeriod.length,
        onTimeCompleted: onTimeInPeriod.length,
        completionRate:
          periodRequirements.length > 0
            ? Math.round(
                (completedInPeriod.length / periodRequirements.length) * 100,
              )
            : 0,
        onTimeRate:
          completedInPeriod.length > 0
            ? Math.round(
                (onTimeInPeriod.length / completedInPeriod.length) * 100,
              )
            : 0,
      };

      res.status(200).json({
        success: true,
        data: metrics,
      });
    } catch (error) {
      logger.error("Error fetching completion metrics:", error);
      res.status(500).json({
        success: false,
        error: {
          message: "Failed to fetch completion metrics",
          code: "COMPLETION_METRICS_ERROR",
        },
      });
    }
  }
}

export const requirementCompletionController =
  new RequirementCompletionController();

File: pma-api-server/src/PMA.Api/Controllers/RequirementCompletionController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using PMA.Core.Entities;
using PMA.Core.Enums;
using PMA.Infrastructure.Data;

namespace PMA.Api.Controllers;

[ApiController]
[Route("api/requirement-completion")]
public class RequirementCompletionController : ApiBaseController
{
    private readonly ApplicationDbContext _context;
    private readonly ILogger<RequirementCompletionController> _logger;

    public RequirementCompletionController(
        ApplicationDbContext context,
        ILogger<RequirementCompletionController> logger)
    {
        _context = context;
        _logger = logger;
    }

    /// <summary>
    /// Get requirement completion analytics including overdue and at-risk requirements
    /// For RequirementCompletionTracker component
    /// </summary>
    [HttpGet("analytics")]
    public async Task<IActionResult> GetCompletionAnalytics(
        [FromQuery] int? analystId = null,
        [FromQuery] int? projectId = null)
    {
        try
        {
            var today = DateTime.UtcNow.Date;

            // Base query for requirements
            var requirementsQuery = _context.ProjectRequirements
                .Include(pr => pr.Project)
                .AsQueryable();

            // Filter by analyst if provided
            if (analystId.HasValue)
            {
                requirementsQuery = requirementsQuery
                    .Where(pr => _context.ProjectAnalysts
                        .Any(pa => pa.ProjectId == pr.ProjectId && pa.AnalystId == analystId.Value));
            }

            // Filter by project if provided
            if (projectId.HasValue)
            {
                requirementsQuery = requirementsQuery.Where(pr => pr.ProjectId == projectId.Value);
            }

            // Get overdue requirements (past deadline and not completed)
            var overdueRequirements = await requirementsQuery
                .Where(pr => pr.ExpectedCompletionDate.HasValue &&
                           pr.ExpectedCompletionDate.Value < today &&
                           pr.Status != RequirementStatusEnum.Completed &&
                           pr.Status != RequirementStatusEnum.Cancelled)
                .Select(pr => new
                {
                    requirementId = pr.Id,
                    requirementTitle = pr.Name,
                    projectId = pr.ProjectId,
                    projectName = pr.Project != null ? pr.Project.ApplicationName : "N/A",
                    priority = pr.Priority,
                    status = pr.Status,
                    ExpectedCompletionDate = pr.ExpectedCompletionDate,
                    daysOverdue = EF.Functions.DateDiffDay(pr.ExpectedCompletionDate!.Value, today),
                    assignedAnalyst = _context.ProjectAnalysts
                        .Where(pa => pa.ProjectId == pr.ProjectId)
                        .Select(pa => pa.Analyst != null ? pa.Analyst.FullName : "N/A")
                        .FirstOrDefault()
                })
                .OrderByDescending(r => r.daysOverdue)
                .ToListAsync();

            // Get at-risk requirements (due within 3 days)
            var atRiskRequirements = await requirementsQuery
                .Where(pr => pr.ExpectedCompletionDate.HasValue &&
                           pr.ExpectedCompletionDate.Value >= today &&
                           pr.ExpectedCompletionDate.Value <= today.AddDays(3) &&
                             pr.Status != RequirementStatusEnum.Completed &&
                           pr.Status != RequirementStatusEnum.Cancelled)
                .Select(pr => new
                {
                    requirementId = pr.Id,
                    requirementTitle = pr.Name,
                    projectId = pr.ProjectId,
                    projectName = pr.Project != null ? pr.Project.ApplicationName : "N/A",
                    priority = pr.Priority,
                    status = pr.Status,
                    ExpectedCompletionDate = pr.ExpectedCompletionDate,
                    daysUntilDeadline = EF.Functions.DateDiffDay(today, pr.ExpectedCompletionDate!.Value),
                    assignedAnalyst = _context.ProjectAnalysts
                        .Where(pa => pa.ProjectId == pr.ProjectId)
                        .Select(pa => pa.Analyst != null ? pa.Analyst.FullName : "N/A")
                        .FirstOrDefault()
                })
                .OrderBy(r => r.daysUntilDeadline)
                .ToListAsync();

            // Get completed requirements count
            var completedCount = await requirementsQuery
                .CountAsync(pr => pr.Status == RequirementStatusEnum.Completed);

            // Get total requirements count
            var totalCount = await requirementsQuery.CountAsync();

            // Get in-progress requirements count
            var inProgressCount = await requirementsQuery
                .CountAsync(pr => pr.Status == RequirementStatusEnum.UnderDevelopment );

            // Get pending requirements count
            var pendingCount = await requirementsQuery
                .CountAsync(pr => pr.Status == RequirementStatusEnum.New || pr.Status == RequirementStatusEnum.ManagerReview);

            // Calculate completion rate
            var completionRate = totalCount > 0 
                ? Math.Round((double)completedCount / totalCount * 100, 2) 
                : 0;

            // Calculate on-time completion rate (completed requirements that were completed before deadline)
            // Note: This assumes we have an actual completion date field, using ExpectedCompletionDate as placeholder
            var onTimeCompletedCount = await requirementsQuery
                .Where(pr => pr.Status == RequirementStatusEnum.Completed &&
                           pr.ExpectedCompletionDate.HasValue)
                .CountAsync();

            var onTimeRate = completedCount > 0 
                ? Math.Round((double)onTimeCompletedCount / completedCount * 100, 2) 
                : 0;

            // Calculate average delay days (simplified - using 0 as placeholder)
            var avgDelayDays = 0.0;

            var analytics = new
            {
                summary = new
                {
                    totalRequirements = totalCount,
                    completedRequirements = completedCount,
                    onTimeCompleted = onTimeCompletedCount,
                    onTimeRate = onTimeRate,
                    avgDelayDays = Math.Round(avgDelayDays, 2)
                },
                overdueItems = overdueRequirements,
                atRiskItems = atRiskRequirements
            };

            return Ok(new
            {
                success = true,
                data = analytics,
                message = "Requirement completion analytics retrieved successfully"
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error occurred while retrieving requirement completion analytics");
            return StatusCode(500, new
            {
                success = false,
                message = "An error occurred while retrieving requirement completion analytics",
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Get requirement completion metrics over time
    /// </summary>
    [HttpGet("metrics")]
    public async Task<IActionResult> GetCompletionMetrics(
        [FromQuery] string period = "month",
        [FromQuery] int? analystId = null,
        [FromQuery] int? projectId = null)
    {
        try
        {
            var today = DateTime.UtcNow.Date;
            DateTime startDate;

            // Determine the period
            switch (period.ToLower())
            {
                case "week":
                    startDate = today.AddDays(-7);
                    break;
                case "quarter":
                    startDate = today.AddMonths(-3);
                    break;
                case "year":
                    startDate = today.AddYears(-1);
                    break;
                case "month":
                default:
                    startDate = today.AddMonths(-1);
                    break;
            }

            // Base query for requirements
            var requirementsQuery = _context.ProjectRequirements
                .Include(pr => pr.Project)
                .AsQueryable();

            // Filter by analyst if provided
            if (analystId.HasValue)
            {
                requirementsQuery = requirementsQuery
                    .Where(pr => _context.ProjectAnalysts
                        .Any(pa => pa.ProjectId == pr.ProjectId && pa.AnalystId == analystId.Value));
            }

            // Filter by project if provided
            if (projectId.HasValue)
            {
                requirementsQuery = requirementsQuery.Where(pr => pr.ProjectId == projectId.Value);
            }

            // Get completion trend data
            var completionTrend = await requirementsQuery
                .Where(pr => pr.ExpectedCompletionDate.HasValue &&
                           pr.ExpectedCompletionDate.Value >= startDate &&
                           pr.ExpectedCompletionDate.Value <= today)
                .GroupBy(pr => pr.ExpectedCompletionDate!.Value.Date)
                .Select(g => new
                {
                    date = g.Key,
                    count = g.Count(),
                    onTimeCount = g.Count(pr => pr.Status == RequirementStatusEnum.Completed)
                })
                .OrderBy(x => x.date)
                .ToListAsync();

            // Get requirements created in period
            var createdInPeriod = await requirementsQuery
                .Where(pr => pr.CreatedAt >= startDate && pr.CreatedAt <= today)
                .CountAsync();

            // Get requirements completed in period
            var completedInPeriod = await requirementsQuery
                .Where(pr => pr.ExpectedCompletionDate.HasValue &&
                           pr.ExpectedCompletionDate.Value >= startDate &&
                           pr.ExpectedCompletionDate.Value <= today)
                .CountAsync();

            // Get on-time completed requirements in period
            var onTimeCompletedInPeriod = await requirementsQuery
                .Where(pr => pr.Status == RequirementStatusEnum.Completed &&
                           pr.ExpectedCompletionDate.HasValue &&
                           pr.ExpectedCompletionDate.Value >= startDate &&
                           pr.ExpectedCompletionDate.Value <= today)
                .CountAsync();

            // Calculate completion rate
            var completionRate = createdInPeriod > 0 
                ? Math.Round((double)completedInPeriod / createdInPeriod * 100, 2) 
                : 0;

            // Calculate on-time rate
            var onTimeRate = completedInPeriod > 0 
                ? Math.Round((double)onTimeCompletedInPeriod / completedInPeriod * 100, 2) 
                : 0;

            var metrics = new
            {
                period,
                startDate,
                endDate = today,
                totalRequirements = createdInPeriod,
                completedRequirements = completedInPeriod,
                onTimeCompleted = onTimeCompletedInPeriod,
                completionRate,
                onTimeRate
            };

            return Ok(new
            {
                success = true,
                data = metrics,
                message = "Requirement completion metrics retrieved successfully"
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error occurred while retrieving requirement completion metrics");
            return StatusCode(500, new
            {
                success = false,
                message = "An error occurred while retrieving requirement completion metrics",
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Get analyst-specific requirement completion performance
    /// </summary>
    [HttpGet("analyst-performance")]
    public async Task<IActionResult> GetAnalystPerformance(
        [FromQuery] int? analystId = null,
        [FromQuery] int top = 10)
    {
        try
        {
            var today = DateTime.UtcNow.Date;

            var analystQuery = _context.Users
                .Where(u => _context.UserRoles.Any(ur => ur.UserId == u.Id && 
                           ur.Role != null && ur.Role.Name.Contains("Analyst")))
                .AsQueryable();

            if (analystId.HasValue)
            {
                analystQuery = analystQuery.Where(u => u.Id == analystId.Value);
            }

            var analystPerformance = await analystQuery
                .Select(u => new
                {
                    analystId = u.Id,
                    analystName = u.FullName,
                    department = u.Department != null ? u.Department.Name : "",
                    
                    totalRequirements = _context.ProjectRequirements
                        .Count(pr => _context.ProjectAnalysts
                            .Any(pa => pa.ProjectId == pr.ProjectId && pa.AnalystId == u.Id)),
                    
                    completedRequirements = _context.ProjectRequirements
                        .Count(pr => pr.Status == RequirementStatusEnum.Completed &&
                                   _context.ProjectAnalysts
                                       .Any(pa => pa.ProjectId == pr.ProjectId && pa.AnalystId == u.Id)),
                    
                    overdueRequirements = _context.ProjectRequirements
                        .Count(pr => pr.ExpectedCompletionDate.HasValue &&
                                   pr.ExpectedCompletionDate.Value < today &&
                                   pr.Status != RequirementStatusEnum.Completed &&
                                   pr.Status != RequirementStatusEnum.Cancelled &&
                                   _context.ProjectAnalysts
                                       .Any(pa => pa.ProjectId == pr.ProjectId && pa.AnalystId == u.Id)),
                    
                    onTimeCompletions = _context.ProjectRequirements
                        .Count(pr => pr.Status == RequirementStatusEnum.Completed &&
                                   pr.ExpectedCompletionDate.HasValue &&
                                   _context.ProjectAnalysts
                                       .Any(pa => pa.ProjectId == pr.ProjectId && pa.AnalystId == u.Id))
                })
                .ToListAsync();

            // Calculate performance metrics
            var performanceWithMetrics = analystPerformance
                .Select(ap => new
                {
                    ap.analystId,
                    ap.analystName,
                    ap.department,
                    ap.totalRequirements,
                    ap.completedRequirements,
                    ap.overdueRequirements,
                    ap.onTimeCompletions,
                    completionRate = ap.totalRequirements > 0 
                        ? Math.Round((double)ap.completedRequirements / ap.totalRequirements * 100, 2) 
                        : 0,
                    onTimeRate = ap.completedRequirements > 0 
                        ? Math.Round((double)ap.onTimeCompletions / ap.completedRequirements * 100, 2) 
                        : 0,
                    overdueRate = ap.totalRequirements > 0 
                        ? Math.Round((double)ap.overdueRequirements / ap.totalRequirements * 100, 2) 
                        : 0
                })
                .OrderByDescending(ap => ap.completionRate)
                .ThenByDescending(ap => ap.onTimeRate)
                .Take(top)
                .ToList();

            return Ok(new
            {
                success = true,
                data = performanceWithMetrics,
                count = performanceWithMetrics.Count,
                message = "Analyst performance retrieved successfully"
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error occurred while retrieving analyst performance");
            return StatusCode(500, new
            {
                success = false,
                message = "An error occurred while retrieving analyst performance",
                error = ex.Message
            });
        }
    }
}
